### ISSUE TITLE
Are tools that update themselves acceptable if they require the user to explicitly run the update command?

### ISSUE THREAD

**Issue by vitorgalvao** [Homebrew/homebrew#46736]
[`imgur-screenshot`](https://github.com/jomo/imgur-screenshot) is [in the main repo](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/imgur-screenshot.rb). It does not have an option to update itself, so [I built one](https://github.com/jomo/imgur-screenshot/pull/52). That PR was left on hold because the author fears [that would make it ineligible for homebrew’s main repo](https://github.com/jomo/imgur-screenshot/pull/52#issuecomment-162322240), citing [your guidelines](https://github.com/Homebrew/homebrew/blob/19cae6daee368186913ffab6194c6299a12cbfde/share/doc/homebrew/Acceptable-Formulae.md#we-dont-like-tools-that-upgrade-themselves).

I do understand why you have that guideline in place, but I feel the change I tried to introduce to `imgur-screenshot` would still make the formula acceptable, since the tool would never update itself without explicit user command/permission, and I feel your restriction is meant to keep in check tools that auto-update (which is understandable).

Am I correct in this assumption? Would `imgur-screenshot` still be eligible for inclusion here with the addition of an update flag? And if not, why not? [`youtube-dl`](https://rg3.github.io/youtube-dl/) has that exact feature and is also [in the main repo](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/youtube-dl.rb). Is that a different case because it’s insanely more popular (not judging the rule if that’s the case, just trying to understand it)?

`imgur-screenshot` consists of a single bash script, nothing more (in case complexity of the tool is a factor for consideration).

Also, pinging @jomo so he also gets notified when there’s a reply.

**Comment by vitorgalvao** [comment]
Also, pinging @jomo so he also gets notified when there’s a reply.

**Comment by UniqMartin** [comment]
Like with most package managers, we prefer to stay in control of what is written to our `Cellar`. That means `brew` should be the only program that changes things in there.

If the software updates itself, Homebrew won't know about this. This can easily lead to user confusion if Homebrew then also offers to upgrade to the same version once the update lands in Homebrew. (And users will be confused even if they explicitly told the program to update itself.) I believe it would leave the user also wondering what the “right” way is to update the software, if there are multiple options. Additionally, it would be bad if, e.g., `imgur-screenshot` was initially installed in `/usr/local/Cellar/imgur-screenshot/1.6.0` but that directory name would no longer reflect the correct version due to a self-update.

That said, every program is free to implement a self-update mechanism, but it would be most desirable if there was an option for `./configure` (or whatever build system is in use) to disable that. In cases like `imgur-screenshot` it might be more practical to resort to patching or offering different “builds” for standalone distribution and inclusion into package managers.

I don't know the back story of `youtube-dl`, but my guess would be very high popularity paired with us not being able to convince upstream to make the self-update optional. (Please correct me if I'm wrong!)

**Comment by DomT4** [comment]
We typically request or implement mechanisms to disable that automatic update process. Keybase prior to recently and `heroku-toolbelt` are two that come to mind immediately.

**Comment by mikemcquaid** [comment]
Yeh, ideally we can configure it out it but as long as it doesn't actually self/auto-update I think we can live with the confusion.

**Comment by jomo** [comment]
> In cases like `imgur-screenshot` it might be more practical to resort to patching or offering different “builds” for standalone distribution and inclusion into package managers.

Thanks for the suggestion, I did it that way.

**Comment by vitorgalvao** [comment]
Thank you all for the fast replies. Closing as solved.
